File: Examples/intel_examples/examples_core_c/fftw3xc/source/dp_plan_dft_r2c_3d.c
/*******************************************************************************
* Copyright 2011-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
! Content:
!       Example of using fftw_plan_dft_r2c_3d function.
!
!****************************************************************************/

#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <float.h>
#include "fftw3.h"

static void init_r(double *x, int *N, int *H);
static int verify_c(fftw_complex *x, int *N, int *H);

int main(void)
{
    /* Sizes of 3D transform */
    int N[3] = {7, 14, 21};

    /* Arbitrary harmonic used to verify FFT */
    int H[3] = {-2, -3, -4};

    /* FFTW plan handle */
    fftw_plan r2c = 0;

    /* Pointer to input/output data */
    double *x = 0;

    /* Execution status */
    int status = 0;


    printf("Example dp_plan_dft_r2c_3d\n");
    printf("3D real-to-complex inplace transform\n");
    printf("Configuration parameters:\n");
    printf(" N = {%d, %d, %d}\n", N[0], N[1], N[2]);
    printf(" H = {%d, %d, %d}\n", H[0], H[1], H[2]);

    printf("Allocate real data array x(%i)\n",2*N[0]*N[1]*(N[2]/2+1));
    x  = fftw_malloc(sizeof(double)*2*N[0]*N[1]*(N[2]/2+1));
    if (0 == x) goto failed;

    printf("Create FFTW plan for 3D r2c inplace FFT\n");
    r2c = fftw_plan_dft_r2c_3d(N[0], N[1], N[2], x, (fftw_complex*)x, FFTW_ESTIMATE);
    if (0 == r2c) goto failed;

    printf("Initialize input for r2c transform\n");
    init_r(x, N, H);

    printf("Compute r2c FFT\n");
    fftw_execute(r2c);

    printf("Verify the result of r2c FFT\n");
    status = verify_c((fftw_complex*)x, N, H);
    if (0 != status) goto failed;

 cleanup:

    printf("Destroy FFTW plan\n");
    fftw_destroy_plan(r2c);

    printf("Free data array\n");
    fftw_free(x);

    printf("TEST %s\n",0==status ? "PASSED" : "FAILED");
    return status;

 failed:
    printf(" ERROR\n");
    status = 1;
    goto cleanup;
}

/* Compute (K*L)%M accurately */
static double moda(int K, int L, int M)
{
    return (double)(((long long)K * L) % M);
}

/* Initialize array x(N) to produce unit peaks at x(H) and x(N-H) */
static void init_r(double *x, int *N, int *H)
{
    double TWOPI = 6.2831853071795864769, phase, factor;
    int n1, n2, n3, S1, S2, S3, index;

    /* Generalized strides for row-major addressing of x */
    S3 = 1;
    S2 = (N[2]/2+1)*2;
    S1 = N[1]*S2;

    if ((N[0]-H[0]%N[0])==0 &&
        (N[1]-H[1]%N[1])==0 &&
        (N[2]-H[2]%N[2])==0)
    {
        factor = 1;
    }
    else
    {
        factor = 2;
    }

    for (n1 = 0; n1 < N[0]; n1++)
    {
        for (n2 = 0; n2 < N[1]; n2++)
        {
            for (n3 = 0; n3 < N[2]; n3++)
            {
                phase  = moda(n1,H[0],N[0]) / N[0];
                phase += moda(n2,H[1],N[1]) / N[1];
                phase += moda(n3,H[2],N[2]) / N[2];
                index = n1*S1 + n2*S2 + n3*S3;
                x[index] = factor * cos( TWOPI * phase ) / (N[0]*N[1]*N[2]);
            }
        }
    }
}

/* Verify that x has unit peak at H */
static int verify_c(fftw_complex *x, int *N, int *H)
{
    double err, errthr, maxerr;
    int n1, n2, n3, S1, S2, S3, index;

    /* Generalized strides for row-major addressing of x */
    S3 = 1;
    S2 = N[2]/2+1;
    S1 = N[1]*S2;

    /*
     * Note, this simple error bound doesn't take into account error of
     * input data
     */
    errthr = 2.5 * log( (double)N[0]*N[1]*N[2] ) / log(2.0) * DBL_EPSILON;
    printf(" Check if err is below errthr %.3lg\n", errthr);

    maxerr = 0;
    for (n1 = 0; n1 < N[0]; n1++)
    {
        for (n2 = 0; n2 < N[1]; n2++)
        {
            for (n3 = 0; n3 < N[2]/2+1; n3++)
            {
                double re_exp = 0.0, im_exp = 0.0, re_got, im_got;

                if ((n1-H[0])%N[0]==0 &&
                    (n2-H[1])%N[1]==0 &&
                    (n3-H[2])%N[2]==0)
                {
                    re_exp = 1;
                }
                else if ((-n1-H[0])%N[0]==0 &&
                         (-n2-H[1])%N[1]==0 &&
                         (-n3-H[2])%N[2]==0)
                {
                    re_exp = 1;
                }

                index = n1*S1 + n2*S2 + n3*S3;
                re_got = x[index][0];
                im_got = x[index][1];
                err  = fabs(re_got - re_exp) + fabs(im_got - im_exp);
                if (err > maxerr) maxerr = err;
                if (!(err < errthr))
                {
                    printf(" x[%i][%i][%i]: ",n1,n2,n3);
                    printf(" expected (%.17lg,%.17lg), ",re_exp,im_exp);
                    printf(" got (%.17lg,%.17lg), ",re_got,im_got);
                    printf(" err %.3lg\n", err);
                    printf(" Verification FAILED\n");
                    return 1;
                }
            }
        }
    }
    printf(" Verified,  maximum error was %.3lg\n", maxerr);
    return 0;
}

File: Examples/intel_examples/examples_core_f/fftw3xf/source/dp_plan_dft_r2c_3d.f90
!===============================================================================
! Copyright 2011-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

! Content:
!       Example of using dfftw_plan_dft_r2c_3d function.
!
!*****************************************************************************

program dp_plan_dft_r2c_3d

  include 'fftw3.f'
  include 'fftw3_mkl.f'

  ! Sizes of 3D transform
  integer, parameter :: N1 = 7
  integer, parameter :: N2 = 13
  integer, parameter :: N3 = 5

  ! Arbitrary harmonic used to verify FFT
  integer :: H1 = 1, H2 = 6, H3 = 1

  ! Need double precision
  integer, parameter :: WP = selected_real_kind(15,307)

  ! Execution status
  integer :: status = 0

  ! Data array for in-place-transform
  real(WP), allocatable :: x_real(:,:,:)

  ! FFTW plans
  integer*8 :: r2c = 0

  print *,"Example dp_plan_dft_r2c_3d"
  print *,"3D real-to-complex in-place transform"
  print *,"Configuration parameters:"
  print '("  N = ["I0","I0","I0"]")', N1, N2, N3
  print '("  H = ["I0","I0","I0"]")', H1, H2, H3

  print *,"Allocate data array for in-place transform"
  allocate ( x_real(2*(INT(N1/2.0)+1), N2, N3), STAT = status)
  if (0 /= status) goto 999

  print *,"Create FFTW real-to-complex plan"
  call dfftw_plan_dft_r2c_3d(r2c, N1, N2, N3, x_real, x_real, FFTW_ESTIMATE)
  if (0 == r2c) goto 999

  print *,"Initialize data for real-to-complex FFT"
  call init(x_real, N1, N2, N3, H1, H2, H3)

  print *,"Compute real-to-complex transform"
  call dfftw_execute(r2c)

  print *,"Verify real-to-complex transform"
  status = verificate(x_real, N1, N2, N3, H1, H2, H3)
  if (0 /= status) goto 999

100 continue

  print *,"Destroy FFTW plan"
  call dfftw_destroy_plan(r2c)

  print *,"Free data array"
  deallocate(x_real)

  if (status == 0) then
    print *, "TEST PASSED"
    call exit(0)
  else
    print *, "TEST FAILED"
    call exit(1)
  end if

999 print '("  Error, status = ",I0)', status
  status = 1
  goto 100

contains

  ! Compute mod(K*L,M) accurately
  pure integer*8 function moda(k,l,m)
    integer, intent(in) :: k,l,m
    integer*8 :: k8
    k8 = k
    moda = mod(k8*l,m)
  end function moda

  ! Initialize x to be inverse of unit peak at H1,H2,H3
  subroutine init(x, N1, N2, N3, H1, H2, H3)
    integer N1, N2, N3, H1, H2, H3
    real(WP) :: x(:,:,:)

    real(WP), parameter:: TWOPI = 6.2831853071795864769_WP
    integer k1,k2,k3

    forall (k1=1:N1, k2=1:N2, k3=1:N3)
      x(k1,k2,k3) = 2 * cos( TWOPI * ( &
        real  (moda(H1,k1-1,N1),WP) / N1 &
        + real(moda(H2,k2-1,N2),WP) / N2 &
        + real(moda(H3,k3-1,N3),WP) / N3)) / (N1*N2*N3)
    end forall
    if (mod(H1,N1)==0 .and. mod(H2,N2)==0 .and. mod(H3,N3)==0) then
      x(1:N1, 1:N2, 1:N3) = x(1:N1, 1:N2, 1:N3) / 2
    end if
  end subroutine init

  ! Verify that x(:,:,:) contains harmonic (H1,H2,H3)
  integer function verificate(xr, N1, N2, N3, H1, H2, H3)
    integer N1, N2, N3, H1, H2, H3
    real(WP) :: xr(:,:,:)

    integer k1, k2, k3
    real(WP) err, errthr, maxerr
    complex(WP) res_exp, res_got

    ! Use Cray pointer to associate complex view with the real array
    ! The first dimension must be divisible by 2
    complex(WP) :: x( size(xr,dim=1)/2, size(xr,dim=2), size(xr,dim=3) )
    pointer (p,x)
    p = loc(xr)

    ! Note, this simple error bound doesn't take into account error of
    ! input data
    errthr = 2.5 * log(real(N1*N2*N3,WP)) / log(2.0_WP) * EPSILON(1.0_WP)
    print '("  Check if err is below errthr " G10.3)', errthr

    maxerr = 0
    do k3 = 1, N3
      do k2 = 1, N2
        do k1 = 1, N1/2+1
          if (mod(k1-1-H1,N1)==0 .and. &
            mod  (k2-1-H2,N2)==0 .and. &
            mod  (k3-1-H3,N3)==0) then
            res_exp = 1.0
          else if (mod(-k1+1-H1,N1)==0 .and. &
            mod       (-k2+1-H2,N2)==0 .and. &
            mod       (-k3+1-H3,N3)==0) then
            res_exp = 1.0
          else
            res_exp = 0.0
          end if
          res_got = x(k1,k2,k3)
          err = abs(res_got - res_exp)
          maxerr = max(err,maxerr)
          if (.not.(err < errthr)) then
            print '("  x("I0","I0","I0"):"$)', k1,k2,k3
            print '(" expected ("G24.17","G24.17"),"$)', res_exp
            print '(" got ("G24.17","G24.17"),"$)', res_got
            print '(" err "G10.3)', err
            print *,"  Verification FAILED"
            verificate = 1
            return
          end if
        end do
      end do
    end do
    print '("  Verified,  maximum error was " G10.3)', maxerr
    verificate = 0
  end function verificate

end program dp_plan_dft_r2c_3d
